### ISSUE TITLE
Generating a dynamic table using pug for Email HTML

### ISSUE THREAD

**Issue by bonellia** [pugjs/pug#3250]
**Pug Version:** 2.0.4

**Node Version:** v13.11.0

## Input JavaScript Values

```js
let baseHTMLOptions = {
        tasks: JSON.stringify(summaryTasks)
    };

const baseHTML = pug.compileFile('src/Utility/Templates/mailBase.pug');
let mailHTML = baseHTML(baseHTMLOptions);
```

## Input Pug

```pug
doctype html
html(lang='en')
  head
    meta(http-equiv='Content-Type', content='text/html; charset=UTF-8')
    meta(name='viewport', content='width=device-width, initial-scale=1')
    meta(http-equiv='X-UA-Compatible', content='IE=edge')
    title
    style(type='text/css').
  body(style='margin:0; padding:0; background-color:#F2F2F2;')
    p The summary of the tasks completed can be seen below:
    center
      table(width='100%', cellspacing='0', cellpadding='0', border='0', bgcolor='#F2F2F2')
        thead
            tr
              th Name
              th Surname
              th Organization
              th Action
        tbody
          each task in taskList
              tr
                td #{task.data.name}
                td #{task.data.surname}
                td #{task.data.organizationUnits[0]}
                td #{task.action}
```

## Expected HTML

```html
<center><table width="100%" cellspacing="0" cellpadding="0" border="0" bgcolor="#F2F2F2">
    <tbody></tbody>
  </table>
  <table style="width:100%" border="1">
    <tbody>
      <tr>
        <th>Name</th>
        <th>Surname</th>
        <th>Organization</th>
        <th>Action</th>
      </tr>
      <tr>
        <td>Mmm</td>
        <td>Ddd</td>
        <td>Hhh </td>
        <td>add</td>
      </tr>
      <tr>
        <td>Aaa</td>
        <td>Ooo</td>
        <td>333</td>
        <td>add</td>
      </tr>
      <tr>
        <td>Aaa</td>
        <td>Bbb</td>
        <td>Ddd</td>
        <td>add</td>
      </tr>
      <tr>
        <td>Aaa</td>
        <td>Ddd</td>
        <td>Vvv </td>
        <td>update</td>
      </tr>
      <tr>
        <td>Caa</td>
        <td>Ccc</td>
        <td>Ooo</td>
        <td>update</td>
      </tr>
      <tr>
        <td>Sss</td>
        <td>Ooo</td>
        <td>Aaa </td>
        <td>update</td>
      </tr>
      <tr>
        <td>Aaa</td>
        <td>Rrr</td>
        <td>Mmm</td>
        <td>remove</td>
      </tr>
      <tr>
        <td>Sss</td>
        <td>Ccc</td>
        <td>Iii</td>
        <td>remove</td>
      </tr>
      <tr>
        <td>Bbb</td>
        <td>Ooo</td>
        <td>Aaa </td>
        <td>remove</td>
      </tr>
    </tbody>
  </table>
</center>
```

## Actual HTML

None, I fail to generate it.

## Additional Comments

Greetings. I do have an issue with the library which doesn't necessarily indicate a bug. My use case is that I use pug to generate e-mail friendly html which cannot have any scripts inside (for whatever reason). What I am trying to achieve is actually quite simple, using "each", dynamically adding table rows. The problem I have is to pass my "tasks" object from my typescript/javascript file to the pug file. I have seen numerous examples on the internet which uses code like app.render() etc. but I don't even use express or anything, I am simply trying to generate HTML using my template. Priorly, I had no trouble generating tables using hard coded options.attribute kinda values that I was able to pass to pug file.
However, take a look:

```js
let addedUserOptions = {
            n1: addTasks[0].data.name,
            sn1: addTasks[0].data.surname,
            o1: addTasks[0].data.organizationUnits[0],
            a1: addTasks[0].action,
            n2: addTasks[1].data.name,
            sn2: addTasks[1].data.surname,
            o2: addTasks[1].data.organizationUnits[0],
            a2: addTasks[1].action,
            n3: addTasks[2].data.name,
            sn3: addTasks[2].data.surname,
            o3: addTasks[2].data.organizationUnits[0],
            a3: addTasks[2].action
        }
```
```pug
table(style='width:100%', border='1')
  tr
    th Name
    th Surname
    th Organization
    th Action
  tr
    td #{n1}
    td #{sn1}
    td #{o1}
    td #{a1}
  tr
    td #{n2}
    td #{sn2}
    td #{o2}
    td #{a2}
  tr
    td #{n3}
    td #{sn3}
    td #{o3}
    td #{a3}
```

This is obviously not how it is meant to be used. I also tried generating tables separately, use them as options.table1: generatedHTML etc. which actually allowed me to render some table, but I had garbage values (probably messed up interpolation.)

I also tried using JSON.stringify and pass it to pug, but without JQuerry or JS within pug, I can't just deserialize and pass these values to my table.

So in short, how can I achieve to pass an object as an options value to use them within pug?
Ideally, I would like something like this:

```js
let baseHTMLOptions = {
        tasks: taskList
    };

const baseHTML = pug.compileFile('src/Utility/Templates/mailBase.pug');
let mailHTML = baseHTML(baseHTMLOptions);
```
```pug
tbody
  each task in #{taskList}
    tr
      td task.data.name
      td task.data.surname
      td task.data.organizationUnits[0]
      td task.action
```

Obviously this won't work, but I am stuck so maybe you guys could provide some tips.

**Comment by bonellia** [comment]
To provide some context, I have a typescript app which collects user data from various sources, then after processing them, it generates tasks for these users: add/update/remove.
By using pug, I am trying to generate some HTML which I intend to send reporting mails which summarizes operations I performed using my app.
I see many use cases for pug on web apps, but my app is not really a web app at all other than interacting some REST/SOAP API endpoints.
An example object I am trying to pass to pug template is as follows:
```json
{
	"167501274": {
		"uniqueId": "167501274",
		"action": "add",
		"data": {
			"uniqueId": "167501274",
			"email": "dedede@gmail.com",
			"name": "Mememe",
			"surname": "DEDEDED",
			"credentials": [],
			"organizationUnits": [
				"HAhahaha "
			],
			"userGroups": [
				"Teknokent Personel"
			],
			"numbers": [
				1088
			],
			"phone": "123142123"
		},
		"dateAdded": 1587814074070
	},
}
```

**Comment by bonellia** [comment]
I think I figured it out. I apologize for the pollution!
Here is the code I used to get my desired result for further reference:

```js
    let summaryTasks = [];

    if (addTasks.length >= 3){
        summaryTasks= [ ...summaryTasks, addTasks[0], addTasks[1], addTasks[2] ];
    }
    if (updateTasks.length >= 3){
        summaryTasks= [ ...summaryTasks, updateTasks[0], updateTasks[1], updateTasks[2] ];
    }
    if (removeTasks.length >= 3){
        summaryTasks= [...summaryTasks, removeTasks[0], removeTasks[1], removeTasks[2] ];
    }
    console.log(summaryTasks);

    let baseHTMLOptions = {
        tasks: summaryTasks
    };

    const baseHTML = pug.compileFile('src/Utility/Templates/mailBase.pug');
    let mailHTML = baseHTML(baseHTMLOptions);
```

```pug
doctype html
html(lang='en')
  head
    meta(http-equiv='Content-Type', content='text/html; charset=UTF-8')
    meta(name='viewport', content='width=device-width, initial-scale=1')
    meta(http-equiv='X-UA-Compatible', content='IE=edge')
    title
    style(type='text/css').
  body(style='margin:0; padding:0; background-color:#F2F2F2;')
    p The summary of the tasks completed can be seen below:
    center
      table(style='width:100%', border='1')
        thead
            tr
              th Name
              th Surname
              th Organization
              th Action
        tbody
          each task in tasks
              tr
                td #{task.data.name}
                td #{task.data.surname}
                td #{task.data.organizationUnits[0]}
                td #{task.action}
```
This generates the following html:
```html
<!DOCTYPE html>
<html lang="en">

  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <title></title>
    <style type="text/css"></style>
  </head>

  <body style="margin:0; padding:0; background-color:#F2F2F2;">
    <p>The summary of the tasks completed can be seen below:</p>
    <center>
      <table style="width:100%" border="1">
        <thead>
          <tr>
            <th>Name</th>
            <th>Surname</th>
            <th>Organization</th>
            <th>Action</th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>Meeeet</td>
            <td>DEmmm</td>
            <td>HAVvvv HAVA ELEKTRONİK SANAYİ VE TİCARET ANONİM </td>
            <td>add</td>
          </tr>
          <tr>
            <td>Assss</td>
            <td>OUuu AKkk</td>
            <td>3Y TEKNOLOJİ YAZILIM DONANIM BİLİŞİM ELEKTRONİK DA</td>
            <td>add</td>
          </tr>
          <tr>
            <td>Allll</td>
            <td>Bayyyy</td>
            <td>DAaaaa TELEKOMÜNİKASYON SİSTEMLERİ ANONİM ŞİRKET</td>
            <td>add</td>
          </tr>
          <tr>
            <td>ABDÜSSAMET</td>
            <td>DERİN</td>
            <td>VLMEDIA YAZILIM TANITIM BİLİŞİM DANIŞMANLIK İÇ VE </td>
            <td>update</td>
          </tr>
          <tr>
            <td>Caa</td>
            <td>Çitttt</td>
            <td>OTtt BILIŞIM ANONIM ŞIRKETI</td>
            <td>update</td>
          </tr>
          <tr>
            <td>Serrr</td>
            <td>Özzz</td>
            <td>AEeee DANIŞMANLIK YAZILIM HİZMETLERİ </td>
            <td>update</td>
          </tr>
          <tr>
            <td>AMmm</td>
            <td>ROooo ZADEH</td>
            <td>Mmmm ARGE TEKNOLOJİLERİ PROJE VE DANIŞMANLIK SAN</td>
            <td>remove</td>
          </tr>
          <tr>
            <td>Serkan</td>
            <td>Canbulat</td>
            <td>İNnnn BİLİŞİM ÇÖZÜMLERİ ANONİM ŞİRKETİ</td>
            <td>remove</td>
          </tr>
          <tr>
            <td>Baaa Diiii</td>
            <td>Özzzz</td>
            <td>AEeee MÜHENDİSLİK DANIŞMANLIK YAZILIM HİZMETLERİ </td>
            <td>remove</td>
          </tr>
        </tbody>
      </table>
    </center>
  </body>

</html>
```
